lang/cs/Org.Apache.REEF.IMRU.Tests/TestTaskManager.cs

```diff
@@ -0,0 +1,702 @@
+            taskManager.RecordFailedTaskDuringSystemShuttingDownState(failedTask2);
+
+            //// master task is killed by driver
+            var masterTask = CreateMockFailedTask(MasterTaskId, TaskManager.TaskKilledByDriver);
+            taskManager.RecordFailedTaskDuringSystemShuttingDownState(masterTask);
+
+            Assert.True(taskManager.AllInFinalState());
+        }
+
+        /// <summary>
+        /// Tests after all the tasks are running, an evaluator fails first, then a task fails with communication error
+        /// </summary>
+        [Fact]
+        public void TestFailedEvaluatorThenFailedTaskAfterTasksAreRunningScenario()
+        {
+            var taskManager = TaskManagerWithTasksAdded();
+            taskManager.SubmitTasks();
+
+            taskManager.RecordRunningTask(CreateMockRunningTask(MasterTaskId));
+            taskManager.RecordRunningTask(CreateMockRunningTask(MapperTaskIdPrefix + 1));
+            taskManager.RecordRunningTask(CreateMockRunningTask(MapperTaskIdPrefix + 2));
+
+            //// Evaluator error
+            var failedTask1 = CreateMockFailedTask(MapperTaskIdPrefix + 1, TaskManager.TaskSystemError);
+            taskManager.RecordTaskFailWhenReceivingFailedEvaluator(CreateMockFailedEvaluator("eId1", failedTask1));
+            Assert.Equal(TaskState.TaskFailedByEvaluatorFailure, taskManager.TaskState(MapperTaskIdPrefix + 1));
+
+            //// system is in shutting down, close all other tasks
+            taskManager.CloseAllRunningTasks(TaskManager.CloseTaskByDriver);
+            Assert.Equal(TaskState.TaskWaitingForClose, taskManager.TaskState(MapperTaskIdPrefix + 2));
+            Assert.Equal(TaskState.TaskWaitingForClose, taskManager.TaskState(MasterTaskId));
+
+            //// Another task may get failed by communication during the shutting down
+            var failedTask2 = CreateMockFailedTask(MapperTaskIdPrefix + 2, TaskManager.TaskGroupCommunicationError);
+            taskManager.RecordFailedTaskDuringSystemShuttingDownState(failedTask2);
+            Assert.Equal(TaskState.TaskClosedByDriver, taskManager.TaskState(MapperTaskIdPrefix + 2));
+
+            //// The task that receives the close from driver now send failed event back to driver
+            var masterTask = CreateMockFailedTask(MasterTaskId, TaskManager.TaskKilledByDriver);
+            taskManager.RecordFailedTaskDuringSystemShuttingDownState(masterTask);
+            Assert.Equal(TaskState.TaskClosedByDriver, taskManager.TaskState(MasterTaskId));
+
+            Assert.True(taskManager.AllInFinalState());
+        }
+
+        /// <summary>
+        /// Tests after all the tasks are running, a task fails first, then an evaluator fails
+        /// </summary>
+        [Fact]
+        public void TestFailedTasksThenFailedEvaluatorAfterAllTasksAreRunningScenario()
+        {
+            var taskManager = TaskManagerWithTasksAdded();
+            taskManager.SubmitTasks();
```

Comment:
Maybe we should create a function to set up `TaskManagerWithTasksSubmitted` and `TaskManagerWithTasksRunning` to reduce total length of these tests, since they use a lot of duplicate setup